File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-no-compile-in-use-frame.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isR3fCallbackStateProperty } from "./utils/is-r3f-callback-state-property.js";
import { resolveR3fCallback } from "./utils/resolve-r3f-callback.js";
import { walkFunctionExecution } from "./utils/walk-function-execution.js";

const THREE_COMPILE_METHOD_NAMES: ReadonlySet<string> = new Set(["compile", "compileAsync"]);

export const r3fNoCompileInUseFrame = defineRule({
  id: "r3f-no-compile-in-use-frame",
  title: "R3F shader precompilation inside useFrame",
  category: "Performance",
  severity: "error",
  recommendation: "Precompile scene materials outside useFrame before they are first displayed",
  create: (context: RuleContext) => {
    const analyzedCallbacks = new Set<EsTreeNode>();
    return {
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        const callback = resolveR3fCallback(node, "useFrame", context.scopes);
        if (!callback || analyzedCallbacks.has(callback)) return;
        analyzedCallbacks.add(callback);
        walkFunctionExecution(callback, context.scopes, (candidate) => {
          if (
            !isNodeOfType(candidate, "CallExpression") ||
            !isNodeOfType(candidate.callee, "MemberExpression") ||
            !THREE_COMPILE_METHOD_NAMES.has(getStaticPropertyName(candidate.callee) ?? "") ||
            (!isR3fCallbackStateProperty(candidate.callee.object, callback, "gl", context.scopes) &&
              !isR3fCallbackStateProperty(
                candidate.callee.object,
                callback,
                "renderer",
                context.scopes,
              ))
          ) {
            return;
          }
          context.report({
            node: candidate,
            message:
              "Renderer shader precompilation runs inside useFrame. Compile once before display instead of rechecking the scene every frame",
          });
        });
      },
    };
  },
});

File: native/oxlint/rules/r3f-no-compile-in-use-frame.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_span::GetSpan;

use crate::{
    context::LintContext,
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const MESSAGE: &str = "Renderer shader precompilation runs inside useFrame. Compile once before display instead of rechecking the scene every frame";

#[derive(Debug, Default, Clone)]
pub struct R3FNoCompileInUseFrame;

impl RuleMeta for R3FNoCompileInUseFrame {
    const NAME: &'static str = "r3f-no-compile-in-use-frame";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Perf;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Disallow renderer shader compilation inside useFrame.",
    };
}

impl Rule for R3FNoCompileInUseFrame {
    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        for node in ctx.nodes().iter() {
            let AstKind::CallExpression(call_expression) = node.kind() else {
                continue;
            };
            for_each_r3f_callback_execution_node(
                call_expression,
                "useFrame",
                ctx,
                |candidate, callback_id, _| {
                    let AstKind::CallExpression(candidate_call) = candidate.kind() else {
                        return;
                    };
                    let Some(member_expression) = candidate_call.callee.as_member_expression()
                    else {
                        return;
                    };
                    if matches!(
                        member_expression.static_property_name(),
                        Some("compile" | "compileAsync")
                    ) && (r3f_callback_state_property_matches(
                        member_expression.object(),
                        callback_id,
                        "gl",
                        ctx,
                    ) || r3f_callback_state_property_matches(
                        member_expression.object(),
                        callback_id,
                        "renderer",
                        ctx,
                    )) {
                        ctx.diagnostic(OxcDiagnostic::error(MESSAGE).with_label(candidate.span()));
                    }
                },
            );
        }
    }
}
